### ISSUE TITLE
Unexpected repetition of case.setup in tests

### ISSUE THREAD

**Issue by billsacks** [ESMCI/cime#692]
In doing some investigation for #690 - but also possibly relevant for #689 - I noticed that various pieces of case setup are done over and over again for tests. I picked a random SMS test (which was NOT rerun), and saw (in CaseStatus):

```
[yslogin2:/glade/p/work/sacks/cesm_code/clm_limitirrig/cime/scripts/tests_1018-1324/SMS_Lm1_D.f10_f10.ICLM45BGCCROP.yellowstone_intel.clm-snowlayers_3_monthly.GC.1018-1324.45.i]$ cat CaseStatus
2016-10-18 13:32:57: <command> ./xmlchange --force CLM_BLDNML_OPTS=-fire_emis --append  </command>
2016-10-18 13:33:00: case.setup starting
2016-10-18 13:33:04: <command> ./xmlchange --force CLM_BLDNML_OPTS=-fire_emis --append  </command>
2016-10-18 13:34:59: case.setup complete
2016-10-18 15:12:20: case.setup starting
2016-10-18 15:12:20: case.setup clean complete
2016-10-18 15:12:22: <command> ./xmlchange --force CLM_BLDNML_OPTS=-fire_emis --append  </command>
2016-10-18 15:12:46: case.setup complete
2016-10-18 15:12:46: case.testbuild starting
2016-10-18 15:13:04: case.testbuild complete
2016-10-18 15:13:54: case.setup starting
2016-10-18 15:13:54: case.setup clean complete
2016-10-18 15:13:55: <command> ./xmlchange --force CLM_BLDNML_OPTS=-fire_emis --append  </command>
2016-10-18 15:14:22: case.setup complete
2016-10-18 15:14:22: case.testbuild starting
2016-10-18 15:15:30: case.testbuild complete
2016-10-18 15:35:31: case.setup starting
2016-10-18 15:35:32: case.setup clean complete
2016-10-18 15:35:34: <command> ./xmlchange --force CLM_BLDNML_OPTS=-fire_emis --append  </command>
2016-10-18 15:35:58: case.setup complete
2016-10-18 15:36:19: Run started
2016-10-18 15:43:07: Run SUCCESSFUL
```

I was initially surprised by the two calls to case.testbuild... but maybe that's expected? (One for sharedlib, one for model??). But even setting that aside, I was surprised to see case.setup and the same xmlchange command repeated so many times for a simple SMS test.

@jedwards4b , @jgfouca : is this expected behavior?

**Comment by billsacks** [comment]
By the way: That xmlchange command comes from the testmod

**Comment by jgfouca** [comment]
Yes, that's odd

**Comment by billsacks** [comment]
Regarding the testmod: In cime4, I'm pretty sure that testmods / user_mods were applied at create_newcase time, ensuring that they were applied once and only once. Now it appears that they are applied in some other phase(s) of case setup. The implications are:

1. The xmlchange commands get executed repeatedly. Often this has no effect, but for these `--append` options, it leads to xml variables like this (for the CLM test `ERP_P60x2_Lm36.f10_f10.ICRUCLM50BGCCROP.yellowstone_intel.clm-clm50cropIrrigMonth_interp`):
 ```xml
<entry id="CLM_BLDNML_OPTS" value="-fire_emis -irrig .true. -fire_emis -irrig .true. -fire_emis -irrig .true. -fire_emis -irrig .true. -fire_emis -irrig .true. -fire_emis -irrig .true. -fire_emis -irrig .true. -fire_emis -irrig .true. -fire_emis -irrig .true. -bgc bgc -crop">
```

2. We have code to check if the contents of a given user_nl file from testmods have already been applied to the user_nl file in the case directory, and if so, don't reapply them. This code wasn't needed in cime4, because testmods were applied once and only once. This new code generally won't cause problems, but it (a) leads to additional complexity, and (b) can be wrong in some edge cases. An example of such an edge case is where testmod 'foo' includes 'bar' which includes 'default'. 'default' sets `x = 17` (among other things), 'bar' sets `x = 42`, then 'foo' sets `x = 17` (as the only thing that it sets). The desired behavior here would be to use the last setting, `x = 17`; but I believe that the current implementation will keep `x = 42`, because it will skip applying the 'foo' testmods.

Is there a reason why cime5 applies testmods in this later stage and repeatedly, rather than following what was done in cime4 and just applying them once at create_newcase time?

**Comment by billsacks** [comment]
I'm realizing there are really two separate issues here:

1. case.setup is called many times for a test

2. the testmods get applied every time case.setup is run

I'm going to rename this issue to be specific to (2), and open a new issue for (1)

**Comment by rljacob** [comment]
@jedwards4b has some thoughts on this
